tagsets/orsee_mysql.php

```diff
@@ -247,5 +247,121 @@ function dump_array($array,$title="",$dolang=true) {
     echo '</TABLE>';
 }
 
+function check_database_upgrade() {
+    global $settings, $system__database_version;
+
+    if (!isset($settings['database_version'])) {
+        $settings['database_version']='2017112700';
+        $done=or_query("INSERT INTO ".table('options')." VALUES (1,'general',NULL,'database_version','".$settings['database_version']."')");
+    }
+    if ($settings['database_version']<$system__database_version) {
+        $done=or_upgrade_database();
+        return $done;
+    } else {
+        return false;
+    }
+}
+
+function upgrade_database_version($new_version) {
+    global $settings;
+    $settings['database_version']=$new_version;
+    $done=orsee_db_save_array(array('option_value'=>$new_version),'options',1,'option_id');
+    if(!$done) log__admin("Database upgrade error. Could not set database version to new version number ".$new_version."!");
+    return $done;
+}
+
+
+function or_upgrade_database() {
+    global $settings, $system__database_version, $system__database_upgrades;
+    
+    if (!isset($system__database_upgrades)) {
+        $system__database_upgrades=array();
+    }
+
+    // check $system__database_upgrades for syntax errors, resort database upgrades by version number
+    $database_upgrades=array(); $i=0; $continue=true;
+    foreach ($system__database_upgrades as $upgr) {
+        $i++;
+        if ($continue) {
+            $error=or_upgrade_check_syntax($upgr);
+            if ($error) {
+                $continue=false;
+                log__admin("Error in upgrade syntax for $system__database_upgrades item nb ".$i.": ".$error." Upgrade aborted.");
+            } else {
+                $upgr['item_nb']=$i;
+                $database_upgrades[$upgr['version']][]=$upgr;
+            }
+        }
+    }
+    if ($continue) {
+        $done=ksort($database_upgrades,SORT_NUMERIC);
+    }
+    
+    // run the updates, stop if there is an error
+    $continue=true;
+    foreach ($database_upgrades as $this_version=>$vupgrades) {
+        if ($this_version>$settings['database_version']) {
+            foreach ($vupgrades as $upgr) {
```

Comment:
May be start transaction here? So if it stops in the middle of series of upgrades, we are left in the state as if no changes were made.

Reply:
The thing is that MyISAM tables in MySQL do not support transactions, yet. And I think we are not ready yet to completely move to InnoDB, since this would require a lot of testing first. So, for the moment, I would not implement transactions yet, and wait until the need arises.

Reply:
It is possible to make it safe and check if DB supports transaction before using them. I may send a patch in a separate PR, IMO updates is ideal case when transaction use is very desirable.

Just for info, Nottingham Uni is using ORSEE with InnoDB since 3.0 upgrade, no DB related issue I am aware off.